wdl-analysis/src/config.rs

```diff
@@ -0,0 +1,247 @@
+//! Configuration for this crate.
+
+use std::sync::Arc;
+
+use tracing::warn;
+use wdl_ast::Severity;
+use wdl_ast::SupportedVersion;
+use wdl_ast::SyntaxNode;
+
+use crate::Rule;
+use crate::SyntaxNodeExt as _;
+use crate::UNNECESSARY_FUNCTION_CALL;
+use crate::UNSUPPORTED_VERSION_FALLBACK;
+use crate::UNUSED_CALL_RULE_ID;
+use crate::UNUSED_DECL_RULE_ID;
+use crate::UNUSED_IMPORT_RULE_ID;
+use crate::UNUSED_INPUT_RULE_ID;
+use crate::rules;
+
+/// Configuration for `wdl-analysis`.
+///
+/// This type is a wrapper around an `Arc`, and so can be cheaply cloned and
+/// sent between threads.
+#[derive(Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub struct Config {
+    /// The actual fields, `Arc`ed up for easy cloning.
+    #[serde(flatten)]
+    inner: Arc<ConfigInner>,
+}
+
+// Custom `Debug` impl for the `Config` wrapper type that simplifies away the
+// arc and the private inner struct
+impl std::fmt::Debug for Config {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Config")
+            .field("diagnostics", &self.inner.diagnostics)
+            .field("fallback_version", &self.inner.fallback_version)
+            .finish()
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(ConfigInner {
+                diagnostics: Default::default(),
+                fallback_version: None,
+            }),
+        }
+    }
+}
+
+impl Config {
+    /// Get this configuration's [`DiagnosticsConfig`].
+    pub fn diagnostics(&self) -> &DiagnosticsConfig {
+        &self.inner.diagnostics
+    }
+
+    /// Get this configuration's fallback version; see
+    /// [`Config::with_fallback_version()`].
+    pub fn fallback_version(&self) -> Option<SupportedVersion> {
+        self.inner.fallback_version
+    }
+
+    /// Return a new configuration with the previous [`DiagnosticsConfig`]
+    /// replaced by the argument.
+    pub fn with_diagnostics(&self, diagnostics: DiagnosticsConfig) -> Self {
```

Comment:
```suggestion
    pub fn with_diagnostics_config(&self, diagnostics: DiagnosticsConfig) -> Self {
```
It's more verbose, but just `with_diagnostics` reads to me like it's expecting a `Vec<Diagnostic>` or some collection of already encountered diagnostics.